feat(security): allow non-root operation and harden service

outlook84 opened the pull request:
Strengthen the security of the `openlist.service` systemd unit by applying a comprehensive set of modern sandboxing and hardening directives.

This change sandboxes the process to limit its potential impact in case of a compromise. Key improvements include:
- Restricting filesystem access (`ProtectSystem`, `PrivateTmp`, `PrivateDevices`)
- Limiting the kernel attack surface with a strict system call filter
- Preventing privilege escalation (`NoNewPrivileges`)
- Dropping unnecessary capabilities (`CapabilityBoundingSet`)
- Isolating the process from devices, IPC, and other system resources
- Custom user/group for systemd service

Additionally, the network dependency is updated to `network-online.target` to ensure the network is fully available before the service starts.

Breaking changes on filesystem access:
- When running as root, `ProtectSystem=true`  makes `/usr`, `/boot` and `/efi` read-only.
- When running as non-root, `ProtectSystem=full`  makes `/usr`, `/etc`,  `/boot` and `/efi` read-only.
- `PrivateTmp=true` gives the service its own isolated /tmp and /var/tmp directories.

outlook84 committed:
feat(systemd): Harden openlist service unit

Strengthen the security of the `openlist.service` systemd unit by applying a comprehensive set of modern sandboxing and hardening directives.

This change sandboxes the process to limit its potential impact in case of a compromise. Key improvements include:
- Restricting filesystem access (`ProtectSystem`, `PrivateTmp`, `PrivateDevices`)
- Limiting the kernel attack surface with a strict system call filter
- Preventing privilege escalation (`NoNewPrivileges`)
- Dropping unnecessary capabilities (`CapabilityBoundingSet`)
- Isolating the process from devices, IPC, and other system resources

Additionally, the network dependency is updated to `network-online.target` to ensure the network is fully available before the service starts.

jyxjjj commented on script/v4.sh:
```diff
@@ -732,14 +732,38 @@ INIT() {
+ProtectKernelLogs=true
+ProtectClock=true
+ProtectHostname=true
+PrivateTmp=disconnected
```
`disconnected` triggers creation of a new tmpfs, which could be unnecessary depending on the context. For standard scenarios, `true` is enough and avoids extra tmpfs allocation. Recommend changing to `true` unless isolation via a new tmpfs is specifically required. Can you explain why you want to use `disconnected`? Is there any related security reason?

jyxjjj commented on script/v4.sh:
```diff
@@ -732,14 +732,38 @@ INIT() {
+SystemCallFilter=~@resources
+SystemCallErrorNumber=EPERM
+SystemCallArchitectures=native
+RemoveIPC=true
```
Although uncertain, but it could potentially affect our upcoming plug-in system.

jyxjjj replied:
Need @dezhishen check this.

jyxjjj commented on script/v4.sh:
```diff
@@ -732,14 +732,38 @@ INIT() {
 Description=OpenList service
-Wants=network.target
-After=network.target network.service
+After=network-online.target nss-lookup.target
```
I don't think NSS is really necessary; the network doesn't even need to be online. That said, making it online could still be a good idea.

jyxjjj commented on script/v4.sh:
```diff
@@ -732,14 +732,38 @@ INIT() {
 WorkingDirectory=$INSTALL_PATH
 ExecStart=$INSTALL_PATH/openlist server
-KillMode=process
+CapabilityBoundingSet=CAP_NET_BIND_SERVICE CAP_DAC_OVERRIDE
+AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_DAC_OVERRIDE
```
I'm not sure `CAP_DAC_OVERRIDE` is a good choice here. Maybe we should keep all ACLs enforced.

jyxjjj replied:
> root user can't

What? Is there any example? This has never happened on my system. Could this be caused by some other parameters being set together?

outlook84 committed:
feat: Allow custom user/group for systemd service

Introduce `SYSTEMD_USER` and `SYSTEMD_GROUP` variables to configure the systemd service user.
Implement `parse_install_args` to allow specifying a custom installation path and user:group via command-line arguments.
Update the systemd service unit to run OpenList under the specified user and group, adjusting capabilities and `ProtectSystem` accordingly.

outlook84 committed:
feat(script): Add permission checks and fix version file path

- Add a `CHECK_PERMISSIONS` function to verify that a non-root user has sufficient permissions to access the installation path before proceeding.
- Make the `.version` file path relative to the installation directory instead of hardcoded.

outlook84 committed:
Refactor(script):  The read (`r`) permission is not needed for parent directory.

Only execute (`x`) permission is required for a service user to traverse a directory path to access its subdirectories. The read (`r`) permission is not necessary for this purpose.

xrgzs reviewed (changes requested):
需要确保用户是否存在

如下为输入了一个不存在的用户和用户组安装，导致无法启动：

```shell
└─$ systemctl status openlist
× openlist.service - OpenList service
     Loaded: loaded (/etc/systemd/system/openlist.service; enabled; preset: disabled)
     Active: failed (Result: exit-code) since Mon 2025-08-25 13:01:19 EDT; 1min 21s ago
   Duration: 5ms
 Invocation: 5bfd5fc580664ba19fe32299ddd55369
    Process: 2630 ExecStart=/opt/openlist/openlist server (code=exited, status=217/USER)
   Main PID: 2630 (code=exited, status=217/USER)
   Mem peak: 1.9M
        CPU: 4ms

Aug 25 13:01:19 kali systemd[1]: Started openlist.service - OpenList service.
Aug 25 13:01:19 kali (openlist)[2630]: openlist.service: Failed to determine user credentials: No such pr>
Aug 25 13:01:19 kali (openlist)[2630]: openlist.service: Failed at step USER spawning /opt/openlist/openl>
Aug 25 13:01:19 kali systemd[1]: openlist.service: Main process exited, code=exited, status=217/USER
Aug 25 13:01:19 kali systemd[1]: openlist.service: Failed with result 'exit-code'.

```

内容：

```shell
└─$ cat /etc/systemd/system/openlist.service
[Unit]
Description=OpenList service
After=network-online.target

[Service]
Type=simple
User=1001
Group=1001
WorkingDirectory=/opt/openlist
ExecStart=/opt/openlist/openlist server
CapabilityBoundingSet=CAP_NET_BIND_SERVICE
AmbientCapabilities=CAP_NET_BIND_SERVICE
NoNewPrivileges=true
ProtectSystem=full
ProtectProc=invisible
ProtectControlGroups=true
ProtectKernelTunables=true
ProtectKernelModules=true
ProtectKernelLogs=true
ProtectClock=true
ProtectHostname=true
PrivateTmp=true
PrivateDevices=true
RestrictNamespaces=true
RestrictRealtime=true
RestrictAddressFamilies=AF_INET AF_INET6 AF_UNIX
RestrictSUIDSGID=true
SystemCallFilter=@system-service
SystemCallFilter=~@privileged
SystemCallFilter=~@resources
SystemCallErrorNumber=EPERM
SystemCallArchitectures=native
RemoveIPC=true
IPAddressDeny=multicast
MemoryDenyWriteExecute=true
LockPersonality=true

[Install]
WantedBy=multi-user.target

```

outlook84 committed:
feat(script): Validate systemd user and group existence

Add a check to verify that the user and group specified for the systemd service exist on the system before proceeding with the installation.

Previously, the script would accept any user:group string, which could lead to a failed service configuration if the user or group was invalid.

This commit introduces a `check_user_group` function that uses `getent` for validation. This check is now integrated into both the command-line argument parsing (`--user`) and the interactive menu, ensuring a more robust installation process by preventing setup with non-existent credentials.

outlook84 committed:
Merge remote-tracking branch 'upstream/main'

ILoveScratch2 committed:
Merge branch 'main' into main

qianbinbin commented:
插个嘴，有的特性需要比较新的 systemd 版本，比如 ProtectProc 需要 247，相当于 Debian 11 或 Ubuntu 22.04

另外有人告诉过我群晖的 systemd 非常古老，不过也许他们可以用 docker？（我不是群晖用户）
